### ISSUE TITLE
Setting TCP Keepalive delay to 0 on Windows 2012r2/2016 results in excessive keepalives sent

### ISSUE THREAD

**Issue by cc-int3** [libuv/libuv#3487]
This can most easily be seen in Node running the following code on windows 2012r2/2016 ( 2019 seems to have argument checking)

```
var Net = require('net');

// Create a server
var srv = Net.createServer(function(s){
  console.log('Connected %j', s.address())
  // Doesn't matter what it does
  s.pipe(s)
});
// Start on some port
srv.listen(1337, function(){
  console.log('Listening on %j', srv.address())
});
// connect to ourself
var s = Net.createConnection({port:1337}, function(){
  console.log('Connected to %j', s.address())
  s.setKeepAlive(true, 0)
});
```
If you look at wireshark (loopback) you'll see:
![ka](https://user-images.githubusercontent.com/16959635/154367993-f648aa6a-a79e-4e79-b91e-d606b95dbf5b.png)

It seems libuv is using an undocumented setsockopt opt, it should just use the windows settings. However, an argument could be made that the current use makes the system more flexible on a per connection basis. In that case, I would like libuv to treat a 0 delay setting as 'use the system settings'

/src/win/tcp.c

```
static int uv__tcp_keepalive(uv_tcp_t* handle, SOCKET socket, int enable, unsigned int delay) {
  if (setsockopt(socket,
                 SOL_SOCKET,
                 SO_KEEPALIVE,
                 (const char*)&enable,
                 sizeof enable) == -1) {
    return WSAGetLastError();
  }

// added && delay > 0
  if (enable && delay > 0 && setsockopt(socket,
                           IPPROTO_TCP,
                           TCP_KEEPALIVE,
                           (const char*)&delay,
                           sizeof delay) == -1) {
    return WSAGetLastError();
  }

  return 0;
}
```

**Comment by bnoordhuis** [comment]
It probably makes more sense to return an error when `delay == 0` because that's what already happens on e.g. Linux because the setsockopt() call fails with EINVAL.
```diff
diff --git a/src/win/tcp.c b/src/win/tcp.c
index 1d9085c9..9a8f7053 100644
--- a/src/win/tcp.c
+++ b/src/win/tcp.c
@@ -58,6 +58,10 @@ static int uv__tcp_nodelay(uv_tcp_t* handle, SOCKET socket, int enable) {
 
 
 static int uv__tcp_keepalive(uv_tcp_t* handle, SOCKET socket, int enable, unsigned int delay) {
+  if (enable)
+    if (delay < 1)
+      return ERROR_INVALID_PARAMETER;
+
   if (setsockopt(socket,
                  SOL_SOCKET,
                  SO_KEEPALIVE,
```

**Comment by cc-int3** [comment]
Note sure how downstream use ( Ie Node ) binds libuv behavior but Node expects:

Setting 0 for initialDelay will leave the value unchanged from the default (or previous) setting
(https://nodejs.org/api/net.html#socketsetkeepaliveenable-initialdelay)

So 0 seems like it should be a no-op, not error

**Comment by bnoordhuis** [comment]
Let me think about this a little more. What you're suggesting sounds okay in principle but it's unclear how calling `uv_tcp_keepalive()` multiple times would work:
```c
uv_tcp_keepalive(handle, 1, 60);
uv_tcp_keepalive(handle, 1, 0); /* Does this reset the delay? How? */
```

**Comment by cc-int3** [comment]
I think in your example the second call would fall under "Setting 0 for initialDelay will leave the value unchanged from the ...previous.. setting" clause in the doc.

I would hope the normal usage would be this :
`uv_tcp_keepalive(handle, 1, 0); /* enable keepalives, use windows settings  for delay*/`
not sure how that lines up with linux behavior

I'll have to test, but i wonder what this behavior would be:
```
uv_tcp_keepalive(handle, 1, 60);
uv_tcp_keepalive(handle, 0, 0); /* disable keepalives */
uv_tcp_keepalive(handle, 1, 0); /* re-enable keepalives...is delay still 60 or windows default */
```

**Comment by cc-int3** [comment]
tested, delay stays at 60. I believe that is acceptable.

```
uv_tcp_keepalive(handle, 1, 60);
uv_tcp_keepalive(handle, 0, 0); /* disable keepalives */
uv_tcp_keepalive(handle, 1, 0); /* re-enable keepalives...delay is still 60 */
```

**Comment by bnoordhuis** [comment]
It's that "unless previously set" I'd qualify as surprising behavior. I also wouldn't want to vouch it works the same on all platforms. That has to be addressed somehow (or at least decided on) before we change it.

@libuv/collaborators you also may want to weigh in here, maybe I'm overthinking this.

**Comment by cc-int3** [comment]
I guess whether the behavior is surprising or not, depends on why someone would try to set the delay multiple times or disable/enable keepalives after enabling them. I can't that think of a (good) reason anyone would do it. Maybe troubleshooting, but that's about it.

**Comment by saghul** [comment]
My 2 cents: I'd expect people to set the desired delay when they enable the feature. It doesn't matter what value is set when disabling, so let's ignore that case for a second. So that leaves the 0 case, which is the weird one. 

~~~
uv_tcp_keepalive(handle, 1, 60);
uv_tcp_keepalive(handle, 0, 0); /* disable keepalives */
uv_tcp_keepalive(handle, 1, 0); /* re-enable keepalives...delay is still 60 */
~~~

I find this surprising behavior TBH. I'm trying to think why would someone set it to 0 instead of the same value you set it earlier? The gains seem ,little for the uncertainty it introduces.

In addition, that "remembering" feature may make someone think this works:

~~~
uv_tcp_keepalive(handle, 1, 60);
uv_tcp_keepalive(handle, 0, 120); /* disable keepalives, but set it to 120 for later */
uv_tcp_keepalive(handle, 1, 0); /* re-enable keepalives...delay is now 120 */
~~~

And it doesn't. I also think it shouldn't.

So, I agree with Ben's proposal here: https://github.com/libuv/libuv/issues/3487#issuecomment-1042688337 we should give an error if delay is < 1.

**Comment by cc-int3** [comment]
I still think you are focusing on the wrong behavior. I think(hope) the 99.44% use case is this:

/* enable keepalives, use system defaults as they are, hopefully, appropriate
    for the local network and an admin wouldn't like to have to set overrides for
    every program and or hope the programmer chose a correct/proper value*/
uv_tcp_keepalive(handle, 1, 0); 

Without the 0 option, you have no way of letting the system decide the initial delay

This code wouldn't do as described, delay is only set when enabling
```
uv_tcp_keepalive(handle, 1, 60);
uv_tcp_keepalive(handle, 0, 120); /* disable keepalives, but set it to 120 for later */
uv_tcp_keepalive(handle, 1, 0); /* re-enable keepalives...delay is now 120 `*/`

```

**Comment by saghul** [comment]
> This code wouldn't do as described, delay is only set when enabling

If 0 uses the previously stored delay, then this limitation makes things really confusing, why can't one store a value for later without enabling it? I'd never do this, I'm just trying to showcase the asymmetry in the API.

**Comment by cc-int3** [comment]
All I'm arguing for is the ability to use the system settings when enabling keepalives[1]. If you have delay < 1 return an error[2] then you've removed an important capability, and you force the developer to pick a delay value, which will be incorrect in all cases[3]. 

It's probably too late now, but the function could have be split in two:

uv_tcp_keepalive(handle, enable);
uv_tcp_keepalive_delay(handle, delay); /* only call if you need a different delay. delay == 0 could be an error, in this case */



[1] And don't forget, the only reason this is an issue is because windows 2012r2/2016 treats 0 as 'send as fast as possible', so the suggested change is only trying to 'emulate' the other platforms tested (2019/ubuntu20.04) "treat 0 as 'leave the value unchanged from the default (or previous) setting'". All the asymmetry currently exists. uv_tcp_keepalive is using an 'undocumented' windows option, TCP_KEEPALIVE, so it's hard to  blame windows for being dumb(in this case).

[2] or specifically delay == 0, my test code used an `int` for delay, not `unsigned int` as in uv_tcp_keepalive. My bad.

[3] Probably.

**Comment by saghul** [comment]
I guess we could also swtich the implementation to https://docs.microsoft.com/en-us/windows/win32/winsock/sio-keepalive-vals and maybe that makes the 0 case behave correctly?

**Comment by cc-int3** [comment]
I'll test it, if it works, I have no problem with that solution.

**Comment by cc-int3** [comment]
SIO_KEEPALIVE_VALS is worse.

`uv_tcp_keepalive(handle, 1, 0)` always sends keepalives as fast as possible ( even on window 10/2019)

**Comment by bnoordhuis** [comment]
@cc-int3 Can you suggest a way forward here? Keep in mind libuv's raison d'être is to abstract away platform differences so the behavior needs to be predictable and consistent across platforms (and edge cases matter.)

**Comment by cc-int3** [comment]
I think my original suggestion gets us to "behavior needs to be ... consistent across platforms". The change brings a few rogue OSes (2012r2/2016) inline with other platforms  (tested: ubuntu 20.04/window 2019) Leaving 0 as a viable initial delay is important to me, as it puts the initial delay into the hands of the sysadmin, not a programmer guessing what a good value should be[*] 

The concerns you and @saghul have is with existing behavior, my change didn't add it, but i think the behavior as it exists now (ignoring old windows) Is covered in the doc. The libuv doc says` ...delay is the initial delay in seconds, ignored when enable is zero.` [**]  However, it doesn't describe what occurs when delay is zero, but that is covered in the node doc `Setting 0 for initialDelay will leave the value unchanged from the default (or previous) setting` That probably should be added to the libuv doc.

So in short, I don't think my suggestion changes any behavior (other than fixing older windows) and the concerns @bnoordhuis and @saghul have regarding behavior are, or should be, covered in the documentation.




[*] Both linux and windows default to 2 hours, which is a far cry from the 5 seconds I've seen set in one project where they decided on a default, due to 0 being non-viable in on older windows. I'm not disparaging them at all, just pointing out the wide swing in various thoughts on the matter. I myself favor 5 minutes for initial delay....what would you pick?

[**] That covers @saghul 's code/concern. 120 should be ignored
`uv_tcp_keepalive(handle, 0, 120); /* disable keepalives, but set it to 120 for later */`

**Comment by saghul** [comment]
I see. If I got it correctly I think I'm +1 here then.

**Comment by cc-int3** [comment]
I added doc to the PR, to hopefully clear up expected behavior
